include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -0,0 +1,571 @@
+#include <seqan3/std/view/subrange.hpp>
+#include <seqan3/std/view/transform.hpp>
+
+#include <seqan3/io/alignment_file/detail.hpp>
+#include <seqan3/io/alignment_file/header.hpp>
+#include <seqan3/io/alignment_file/sam_tag_dictionary.hpp>
+
+namespace seqan3
+{
+/*!\brief       The sam format.
+ * \implements  alignment_file_format_concept
+ * \ingroup     alignment
+ *
+ * \details
+ *
+ * ### Introduction
+ *
+ * sam is the de-facto-standard for alignment storage in bionformatics. See the
+ * [article on wikipedia](https://en.wikipedia.org/wiki/sam_format) for a an in-depth description of the format.
+ *
+ * ### Fields
+ *
+ * The sam format provides the fields seqan3::field::SEQ and seqan3::field::ID. Both fields are required when writing.
+ *
+ * ### Implementation notes
+ *
+ * When reading the ID-line the identifier (either `;` or `>`) and any blank characters before the actual ID are
+ * stripped.
+ *
+ * This implementation supports the following less known and optional features of the format:
+ *
+ *   * ID lines beginning with `;` instead of `>`
+ *   * line breaks and other whitespace characters in any part of the alignment
+ *   * character counts within the alignment (they are simply ignored)
+ *
+ * The following optional features are currently **not supported:**
+ *
+ *   * Multiple comment lines (starting with either `;` or `>`), only one ID line before the alignment line is accepted
+ *
+ */
+class alignment_file_format_sam
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \brief Rule of five explicitly defaulted.
+     * \{
+     */
+    alignment_file_format_sam() = default;
+    alignment_file_format_sam(alignment_file_format_sam const &) = delete;
+    alignment_file_format_sam & operator=(alignment_file_format_sam const &) = delete;
+    alignment_file_format_sam(alignment_file_format_sam &&) = default;
+    alignment_file_format_sam & operator=(alignment_file_format_sam &&) = default;
+    //!\}
+
+    //!\brief The valid file extensions for this format; note that you can modify this value.
+    static inline std::vector<std::string> file_extensions
+    {
+        { "sam" },
+    };
+
+    //!\copydoc alignment_file_output_format_concept::write
+    template <typename stream_type,
+              typename header_type,
+              typename seq_type,
+              typename id_type,
+              typename offset_type,
+              typename ref_seq_type,
+              typename ref_id_type,
+              typename ref_offset_type,
+              typename align_type,
+              typename flag_type,
+              typename mapq_type,
+              typename qual_type,
+              typename mate_type,
+              typename tag_dict_type,
+              typename evalue_type,
+              typename bit_score_type>
+    void write(stream_type                      &  stream,
+               alignment_file_out_options const &  options,
+               header_type                      && header,
```

Comment:
because we decided on this in another discussion...
the main point was that otherwise the direct range-based functionality does not work. E. g. `file_out = file_in;`

Reply:
This is especially important considering BAM files: Here, the header is required because the reference names are stored as numbers and the ref_dict is the only way to restore the names. When reading a BAM file without reading the header first, accessing the ref_name will result in the wrong behaviour.

Reply:
> because we decided on this in another discussion...

Yes, I remember, thanks for reminding me! Still I would make the header a member of the file only include a pointer to it in the record. I would then just make a check directly at the beginning of the first write call so that it is written.  But there shouldn't be an extra write-call with std::ignores just for the header (I don't know if I maybe misunderstood that from your comments)

Reply:
Ah now I see what you meant! Thanks for clarifying.